unic/utils/src/macros.rs

```diff
@@ -0,0 +1,307 @@
+/// Macro for declaring a character property.
+///
+/// # Syntax
+///
+/// ```
+/// # #[macro_use] extern crate unic_utils;
+/// # fn main() {}
+/// char_property! {
+///     /// Zero or more attributes
+///     pub enum PropertyName {
+///         /// Exactly one attribute
+///         VariantName {
+///             abbr => AbbrName,
+///             long => Long_Name,
+///             display => "&'static str that is a nicer presentation of the name",
+///         }
+///
+///         /// All annotations on the variant are optional
+///         Variant2 {}
+///     }
+///
+///     /// Zero or more attributes
+///     pub mod abbr_names;
+/// }
+/// ```
+///
+/// # Effect
+///
+/// - Implements `CharProperty` with the `abbr` and `long` presented in the appropriate method
+/// - Implements `FromStr` accepting the variant name or the abbr or long forms
+/// - Implements `Display` using the given string, falling back when not provided on
+///   the long name, the short name, and the variant name, in that order
+/// - Populates the module `abbr_names` with `pub use` bindings of variants to their short names
+/// - Maintains all documentation comments and other `#[attributes]` as would be expected
+///
+/// # Limitations
+///
+/// Due to [rust-lang/rust/#24189](https://github.com/rust-lang/rust/issues/24189), (fixed in
+/// [rust-lang/rust/#42913](https://github.com/rust-lang/rust/pull/42913), landing in 1.20),
+/// exactly one attribute line must be used on each variant. On 1.20 or higher, one or more may
+/// be used, and the restriction can be relaxed back the intended zero or more by replacing
+/// `$(#[$variant_meta:meta])+` with `$(#[$variant_meta:meta])*`, and
+/// `$(#[$variant_meta])+` with `$(#[$variant_meta])*`, and
+/// `$(#[$ident_meta:meta])+` with `$(#[$ident_meta:meta])*` and
+/// `$(#[$ident_meta])+` with `$(#[$ident_meta])*`, and
+/// `$(#[$rest_meta:meta])+` with `$(#[$rest_meta:meta])*`, and
+/// `$(#[$rest_meta])+` with `$(#[$rest_meta])*`, and
+/// `$(#[$queue_meta:meta])+` with `$(#[$queue_meta:meta])*`, and
+/// `$(#[$queue_meta])+` with `$(#[$queue_meta])*`
+// TODO: Once adopting 1.20, fix the macro to work with zero attributes on variants (see above)
+#[macro_export]
+macro_rules! char_property {
+    (
+        $(#[$name_meta:meta])* pub enum $name:ident {
```

Comment:
The problem with that is `pub` doesn't match `#[$:meta]`. The proper answer would be `$:vis` (rust-lang/rust#41022) but that's still unstable or doing it manually, which is slightly overkill.